tools/meson-render-jinja2.py

```diff
@@ -8,33 +8,38 @@
 
 import jinja2
 
+
 def parse_config_h(filename):
     # Parse config.h file generated by meson.
     ans = {}
-    for line in open(filename):
-        m = re.match(r'#define\s+(\w+)\s+(.*)', line)
-        if not m:
-            continue
-        a, b = m.groups()
-        if b and b[0] in '0123456789"':
-            b = ast.literal_eval(b)
-        ans[a] = b
-    return ans
+    with open(filename, encoding='utf-8') as f:
```

Comment:
I see, makes sense, thanks for the explanation.

Reply:
> The idea that a context manager should be used for files open for reading is a cargo-cult that has no technical justification.

Code that isn't entirely sloppy tends to be run with warning enabled from time to time and with warnings enabled python complains about unclosed files a lot so if nothing else it's used to make it easier to turn on runtime checks. Then again I'm not sure if @mrc0mmand is planning to roll out that on the CI :-)
```python
Exception ignored in: <_io.FileIO name='/dev/fd/63' mode='rb' closefd=True>
Traceback (most recent call last):
  File "tools/analyze-dump-sort.py", line 69, in <module>
    one = sort_dump(open(opts.one))
          ^^^^^^^^^^^^^^^^^^^^^^^^^
ResourceWarning: unclosed file <_io.TextIOWrapper name='/dev/fd/63' mode='r' encoding='UTF-8'>
Exception ignored in: <_io.FileIO name='/dev/fd/62' mode='rb' closefd=True>
Traceback (most recent call last):
  File "tools/analyze-dump-sort.py", line 71, in <module>
    two = sort_dump(open(opts.two))
          ^^^^^^^^^^^^^^^^^^^^^^^^^
ResourceWarning: unclosed file <_io.TextIOWrapper name='/dev/fd/62' mode='r' encoding='UTF-8'>
Exception ignored in: <function Popen.__del__ at 0x7f516c6c7100>
Traceback (most recent call last):
  File "/usr/lib64/python3.11/subprocess.py", line 1127, in __del__
    _warn("subprocess %s is still running" % self.pid,
ResourceWarning: subprocess 2400 is still running
Exception ignored in: <_io.FileIO name='/tmp/tmp427dcmr_' mode='wb' closefd=True>
ResourceWarning: unclosed file <_io.TextIOWrapper name='/tmp/tmp427dcmr_' mode='wt' encoding='UTF-8'>
Exception ignored in: <_io.FileIO name='/tmp/tmpissj47qm' mode='wb' closefd=True>
ResourceWarning: unclosed file <_io.TextIOWrapper name='/tmp/tmpissj47qm' mode='wt' encoding='UTF-8'>
```

Reply:
> that warning is part of the same misunderstanding

FWIW It was discussed in https://github.com/python/cpython/issues/54302 and there it was decided that it would be better to emit that warning anyway (and `open.read()` combined with garbage collection was outvoted there) but yeah those warnings can be suppressed of course.

Reply:
That was 13 years ago ;)  I skimmed the discussion, and even back then, the motivation was Windows, because of open file exclusivity and strange limits. In the thread resource loops are mentioned. But even 13 years ago this wouldn't have been an issue on other OSes. And its become even more so: for us now fds can be treated like memory — an essentially unlimited resource, up to the memory limits of the machine. The build tools that we write don't have to care about anything but Linux.

Reply:
I'm not saying I disagree :-)

My point is that the standard tooling intentionally flags that to cover code that can be run (or copy-pasted) in all sorts of environments including, say, PyPi where files aren't closed when they go out of scope. It isn't applicable to short-lived scripts glued together with duct tape of course but even those scripts are (hopefully) run with warnings enabled from time to time and to shut Python up they can either go with the flow or disable that explicitly. Someone has to type a few keystrokes somewhere anyway :-) The third option is not to run anything with warnings enabled and everyone is happy :-)